components/l10n/common/locale_subtag_parser_util.cc

```diff
@@ -0,0 +1,104 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/l10n/common/locale_subtag_parser_util.h"
+
+#include "absl/types/optional.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "brave/components/l10n/common/locale_subtag_info.h"
+
+namespace brave_l10n {
+
+namespace {
+
+constexpr char kHyphenSeparator = '-';
+constexpr char kUnderscoreSeparator = '_';
+constexpr char kCodeSetSeparator = '.';
+constexpr char kVariantSeparator = '@';
+
+std::string NormalizeLocale(const std::string& locale) {
+  // Calculate length of locale excluding charset and variant.
+  const std::string::size_type pos = locale.find_first_of(base::StrCat(
+      {std::string{kCodeSetSeparator}, std::string{kVariantSeparator}}));
+
+  const size_t length = (pos == std::string::npos) ? locale.length() : pos;
+
+  // Normalize language, script, and country separators.
+  std::string normalized_locale{locale};
+  for (size_t i = 0; i < length; i++) {
+    if (normalized_locale[i] == kHyphenSeparator) {
+      normalized_locale[i] = kUnderscoreSeparator;
+    }
+  }
+
+  return normalized_locale;
+}
+
+}  // namespace
+
+LocaleSubtagInfo ParseLocaleSubtags(const std::string& locale) {
+  static absl::optional<LocaleSubtagInfo> cached_locale_subtag;
+
+  static std::string last_locale;
+
+  if (cached_locale_subtag && last_locale == locale) {
+    return *cached_locale_subtag;
+  }
+
+  last_locale = locale;
+
+  LocaleSubtagInfo locale_subtag;
+
+  if (locale.empty()) {
+    cached_locale_subtag = locale_subtag;
+    return locale_subtag;
+  }
+
+  std::string str = NormalizeLocale(locale);
+
+  // Parse variant.
+  std::string::size_type pos = str.find(kVariantSeparator);
+  if (pos != std::string::npos) {
+    locale_subtag.variant = {str, pos + 1};
+    str = {str, 0, pos};
```

Comment:
it is better to use constructor syntax in such places: https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++-dos-and-donts.md#variable-initialization

Reply:
makes the code more verbose, but rules are rules! :-)